environments/gravity.py

```diff
@@ -0,0 +1,213 @@
+                    brush = self.colors[n]
+                    color_term = np.exp(-(((I - q[n, 0, t])**2 +
+                                           (J - q[n, 1, t])**2) /
+                                          (self.mass[n]**2))**4)
+                    if brush == 'y':
+                        vid[t, :, :, 0] += color_term
+                        vid[t, :, :, 1] += color_term
+                        vid[t, :, :, 2] -= color_term
+                    elif brush == 'r':
+                        vid[t, :, :, 0] += color_term
+                        vid[t, :, :, 1] -= color_term
+                        vid[t, :, :, 2] -= color_term
+                    elif brush == 'g':
+                        vid[t, :, :, 0] -= color_term
+                        vid[t, :, :, 1] += color_term
+                        vid[t, :, :, 2] -= color_term
+                    elif brush == 'b':
+                        vid[t, :, :, 0] -= color_term
+                        vid[t, :, :, 1] -= color_term
+                        vid[t, :, :, 2] += color_term
+                    elif brush == 'c':
+                        vid[t, :, :, 0] -= color_term
+                        vid[t, :, :, 1] += color_term
+                        vid[t, :, :, 2] += color_term
+                    elif brush == 'p':
+                        vid[t, :, :, 0] += color_term
+                        vid[t, :, :, 1] -= color_term
+                        vid[t, :, :, 2] += color_term
+                    elif brush == 'w':
+                        vid[t, :, :, 0] += color_term
+                        vid[t, :, :, 1] += color_term
+                        vid[t, :, :, 2] += color_term
+                    vid[t][vid[t] < 0] = 0
+            else:
+                for i in range(self.n_objects):
+                    vid[t, :, :, 0] += np.exp(-(((I - q[i, 0, t])**2 +
+                                                 (J - q[i, 1, t])**2) /
+                                                (self.mass[i]**2))**4)
+            vid[t][vid[t] > 1] = 1
+            vid[t][vid[t] < 0] = 0
+
+        return vid
+
+    def _sample_init_conditions(self, radius_bound):
+        """Samples random initial conditions for the environment
+        Args:
+            radius_bound (float, float): Radius lower and upper bound of the phase state sampling.
+        """
+        radius_lb, radius_ub = radius_bound
+        radius = np.random.rand()*(radius_ub - radius_lb) + radius_lb
+        states = np.zeros((2, self.n_objects, 2))
+        # first position
+        pos = np.random.rand(2)*2. - 1
+        pos = (pos/np.sqrt((pos**2).sum()))*radius
+
+        # velocity that yields a circular orbit
+        if self.n_objects == 2:
+            factor = 2
+        else:
+            factor = np.sqrt(np.sin(np.pi/3)/(2*np.cos(np.pi/6)**2))
+        vel = np.flipud(pos)/(factor*radius**1.5)
+        vel[0] *= -1
+        states[0, 0, :] = pos
+        states[1, 0, :] = vel
```

Comment:
The problem with the n-body system is that the trajectories are too chaotic for n>2. For the implementation of the initial conditions, I followed the following code
[https://github.com/greydanus/hamiltonian-nn/blob/master/experiment-3body/data.py](url)
I have not checked whether this conditions lead to very constrained trajectories. The idea was to follow the hamiltonian neural network data.

We could try setting less rigid initial conditions but it would be too difficult to find stable trajectories. The previous solution ensures that the trajectory will be stable for 2 and 3 objects. I generalized the approach to make it work for >3 objects, but stability is not guaranteed in these scenarios.

I don't think that discarding sequences is an option if we want to train our models and generate data at the same time.

Reply:
Ok if they do that in HNN I suppose it is good enough. Let's keep it like this for now